Sgp30 sensor improvements

huubeikens opened the pull request:
## Description:

Updated the SGP30 sensor with configuration options to persistently store the eCO2/TVOC baselines so that it survives a power cycle. It removes the need to recompile and OTA the configuration file when the baselines have been determined after the calibration period (of 12 hours). 

As a convenience, the baseline values can also be configured as sensors so that they are published to Home assistant. 

Last, the update interval default is set to 1 second which is advised by the datasheet.

Please note that the amount of writes to the NVS (non-volatile-storage) is kept to a mininum to prevent wear and tear. 

**Related issue (if applicable):** fixes <link to issue>
#673
#668

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#982

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
Added option to store the baselines for SGP30 sensor. Also created sensors for the baselines and warning when update interval is not accoring to SGP30 datasheet.

Commit:
Change default update interval to 1 second.

Commit:
Set store_baselines default to yes

Commit:
Set some defines back to their original values.

Commit:
Merge branch 'dev' of https://github.com/esphome/esphome into sgp30-sensor-improvements

Commit:
Added mechanisme to prevent too many flash writes

kbx81 commented:
I am running this branch on four sensor boards I have around my place -- so far, so good! I'll report back in a few days with an update. 😄

huubeikens commented:
Thanks for testing!

kbx81 commented:
So far, so good! It's been almost two weeks and still going strong...

huubeikens committed:
Merge branch 'dev' into sgp30-sensor-improvements

huubeikens committed:
Updated to defined constants

Updated to defined constants

huubeikens committed:
Moved const to generic const.py file

css811 and sgp30 components used the same configuration constants. Moved them to const.py

huubeikens committed:
Updated conf constants

huubeikens committed:
Put file through black formatter

huubeikens committed:
@huubeikens Put file through black formatter

glmnet committed:
Merge branch 'dev' into sgp30-sensor-improvements

huubeikens committed:
Merge branch 'dev' into sgp30-sensor-improvements

darkson95 commented:
Thanks for the cool work! This solves exactly the problems I would like to eliminate. 😎
Without building up pressure, I'd like to know when this PR goes live, as I'm having trouble reading the baseline in time.

huubeikens committed:
Merge branch 'dev' into sgp30-sensor-improvements

huubeikens commented:
Thanks! I've been testing it for a few months now without any issues. 

@OttoWinter : could you please merge this PR into the master branch? Much thanks in advance.

agners commented:
There are some linter errors it seems. Otherwise, from my point of view the PR looks good.

agners commented on esphome/components/sgp30/sgp30.h:
```diff
@@ -3,16 +3,25 @@
+struct SGP30Baselines {
+  uint16_t eco2;
+  uint16_t tvoc;
+} PACKED;  // NOLINT
```
I think this `NOLINT` is not required here.

huubeikens replied:
You're right. I've removed it. I've also compiled it again towards esphome:dev and solved a yield error I got. 
See new commits.
This PR should now be ready for the merge into esphome:dev.

huubeikens committed:
Removed lint errors

huubeikens committed:
Removed yield command

kbx81 reviewed (approved):
This has been running on several of my custom sensor boards for quite a while now and it appears to be stable and is working as expected. Nothing jumps out at me code-wise...probably good to merge.

huubeikens commented:
I see that all checks have passed. Please let me know if I need to do something more to get this change merged.

darkson95 commented:
A few days have now passed since this PR is ready to merge. What is still missing here? Is there anything I can assist with so that the merge is accelerated?
